File: BaekJoon/Graph/BFS/벽부수고이동하기_2206.cpp
//
//  벽부수고이동하기_2206.cpp
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/03/01.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: - 벽부수고이동하기
// MARK: BFS C++ 문제풀이

#if 0
#include <iostream>
#include <queue>
#include <vector>
#define MAX 2e9
using namespace std;

typedef pair<pair<int,int>, int> TPair;
vector<vector<int>> SV(1001, vector<int>(1001,0));
int dist[1001][1001][2];

int N,M;
int tx[4] = {0,0,-1,1};
int ty[4] = {1,-1,0,0};

int BFS() {
    queue<TPair> Q;
    Q.push({{0,0},0});
    dist[0][0][0] = 1;
    
    while(!Q.empty()) {
        TPair node = Q.front();
        Q.pop();
        int nx = node.first.first;
        int ny = node.first.second;
        int broken = node.second;
        
        if(nx==N-1 && ny==M-1) {
            return dist[nx][ny][broken];
        }
        
        for(int i=0; i<4; i++) {
            int mx = nx + tx[i];
            int my = ny + ty[i];
            if(mx < 0 || my < 0 || mx >= N || my >= M) continue;
            if(dist[mx][my][broken] != 0) continue;
            
            if(SV[mx][my] == 0) {
                dist[mx][my][broken] = dist[nx][ny][broken] + 1;
                Q.push({{mx,my},broken});
            } else if(broken == 0) {
                dist[mx][my][1] = dist[nx][ny][broken] + 1;
                Q.push({{mx,my},1});
            }
        }
    }
    return -1;
}

int main() {
    cin>>N>>M;
    
    for(int i=0; i<N; i++) {
        for(int j=0; j<M; j++) {
            scanf("%1d",&SV[i][j]);
        }
    }
    
    if(N==1 && M==1) return !printf("%d\n",1);
    
    printf("%d\n", BFS());
    
    return 0;
}
#endif

File: BaekJoon/Graph/BFS/벽부수고이동하기_2206.swift
//
//  벽부수고이동하기_2206.swift
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/03/01.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: TLE 답안

/*
import Foundation

let arr = readLine()!.split(separator: " ").map { Int($0)! }
let N = arr[0], M = arr[1]
let tx = [0,0,-1,1]
let ty = [1,-1,0,0]

var G = [[Character]](repeating: [Character](), count: N)
var C = [[[Int]]](repeating: [[Int]](repeating: [Int](repeating: 0, count: 2), count: 1001), count: 1001)

for i in 0..<N {
    G[i] = Array(readLine()!)
}

if N == 1 && M == 1 { print(1); exit(0) }

func BFS() -> Int {
    var Q = [((Int,Int), Int)]()
    Q.append(((0,0),0))
    C[0][0][0] = 1
    
    while !Q.isEmpty {
        let node = Q.first!
        Q.removeFirst()
        let nx = node.0.0
        let ny = node.0.1
        let broken = node.1
        for i in 0..<4 {
            let mx = nx + tx[i]
            let my = ny + ty[i]
            if mx < 0 || my < 0 || mx >= N || my >= M { continue }
            if C[mx][my][broken] > 0 { continue }
            
            if G[mx][my] == "0" {
                C[mx][my][broken] = C[nx][ny][broken] + 1
                Q.append(((mx,my),broken))
            } else if broken == 0 {
                C[mx][my][1] = C[nx][ny][broken] + 1
                Q.append(((mx,my),1))
            }
            
            if mx == N-1 && my == M-1 { return C[mx][my][broken] }
        }
    }
    return -1
}

print(BFS())
*/
